Volume type standard

josephineSei opened the pull request:
This creates the new standard for volume types as mentioned in ticket #265

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+## DEFAULT volume types
+
+There is always a default volume type defined in an OpenStack deployment. The SCS will does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment.
```
```suggestion
There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment.
```

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+Availability Zones are not necessarily used by every deployer of a cloud. The SCS does not want do make it necessary to have AZs for a compatible deployment. But we want to encourage deployer who use multiple AZs to read the following:
+
+While there can be multiple Compute AZs, there can also be multiple Storage AZs for Volumes. And it might be not quite confusing for users, which volumes can be used in which AZs. To make it even further complicated, there are backends like ceph, which can provide volumes for multiple compute AZs just with some Nova configuration. Therefore we would encourage to use either the property of volume  ...
```
```suggestion
While there can be multiple Compute AZs, there can also be multiple Storage AZs for Volumes. And it might be quite confusing for users, which volumes can be used in which AZs. To make it even further complicated, there are backends like ceph, which can provide volumes for multiple compute AZs just with some Nova configuration. Therefore we would encourage to use either the property of volume types OR the description of the volume types to describe, in which AZs the volumes based on this type can be used.
```

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+### Multiattach
+
+A few backends enable providers to allow volumes to be attached to mulitple VMs simultaneously. The OpenStack API can be used to enable the usage of multiattach through the volume type. The property is also automatically shown to users. If a deployer want to use this feature they have to do it in this way. So users will always see whether a volume type can be used for mulitattach volumes or not.  ...
```
```suggestion
A few backends enable providers to allow volumes to be attached to mulitple VMs simultaneously. The OpenStack API can be used to enable the usage of multiattach through the volume type. The property is also automatically shown to users. If a deployer wants to use this feature they have to do it in this way. So users will always see whether a volume type can be used for mulitattach volumes or not. Nevertheless due to other problems, that might occur when using the multiattach feature, the SCS standard will only recommend|require(?) this as OPTIONAL.
```

which problems exactly? maybe add a link describing them if it's too long for the standard document.

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+### Volume QoS
+
+Quality of Service for volumes can be defined in volume qos objects. While admins can use these directly on volumes, a user is not able to see these object. Instead there can be ONE volume qos object associated to a volume type, which is then be used on all volumes created from that volume type. Through this indirection, functioning volume types without these associated qos objects and this featu ...
```
```suggestion
Quality of Service for volumes can be defined in volume qos objects. While admins can use these directly on volumes, a user is not able to see these objects. Instead there can be ONE volume qos object associated to a volume type, which is then be used on all volumes created from that volume type. Through this indirection, functioning volume types without these associated qos objects and this feature not being heavily used, the SCS will currently only state this as an OPTIONAL recommended(?) volume type feature.
```

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+Quality of Service for volumes can be defined in volume qos objects. While admins can use these directly on volumes, a user is not able to see these object. Instead there can be ONE volume qos object associated to a volume type, which is then be used on all volumes created from that volume type. Through this indirection, functioning volume types without these associated qos objects and this featu ...
+
+To make users aware that a volume type includes specific qos options, we reccomend to write it into the description of a volume type, as any association to a volume qos object cannot be seen by normal users:
```
```suggestion
To make users aware that a volume type includes specific qos options, we recommend to write it into the description of a volume type, as any association to a volume qos object cannot be seen by normal users:
```

It is not clear from the text if we recommend this exact naming from the example or if the naming can be arbitrary chosen.

artificial-intelligence reviewed (changes requested):
I'll do another review later today.

Thanks for all the work you put into this!

josephineSei commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+| Aspect | Standardize? | Discoverability | comments |
+| ---- | ---- | ---- | ------ |
+| encryption | **Recommended** | admin only | Upstream work for extra_spec: encrypted=True/False needed |
```
as soon as there is more than just some discussions at a ptg, i will add a link to a spec or patch.

josephineSei commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+| id                 | d63307fb-167a-4aa0-9066-66595ea9fb21                                                                                                                         |
+| is_public          | True                                                                                                                                                         |
+| name               | LUKS                                                                                                                                                         |
+| properties         | encrypted='true'                                                                                                                                             |
```
I renamed it in the test above.

josephineSei replied:
could you resolve the change request? I edited the document, but not in the table here.

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+## Motivation
+
+We want to standardize a few varieties of volume types. While a user can choose simple things like size when creating a volume, Volume Types define a few broader aspects of volume. Encryption of volumes for example is solely decided by the volume type. While the option with which the volume will be replicated is a mix between definition in the volume type and backend specific configuration, but i ...
```
nit:
```suggestion
We want to standardize a few varieties of volume types. While a user can choose simple things like size when creating a volume, Volume Types define a few broader aspects of a volume. Encryption of volumes for example is solely decided by the volume type. While the option with which the volume will be replicated is a mix between definition in the volume type and backend specific configuration, but it's visibility can only be reached in the volume type. In the following part, we want to state, which varieties of volume types are REQUIRED, RECOMMENDED or only OPTIONAL within an SCS-compatible deployment.
```

artificial-intelligence reviewed (approved):
LGTM

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+| Aspect | Standardize? | Discoverability | comments |
+| ---- | ---- | ---- | ------ |
+| encryption | **Recommended** | admin only | Upstream work for extra_spec: encrypted=True/False needed |
```
Why is should the aspect, if a volume is encrypted or not, only be discoverable by role `admin`?

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+## DEFAULT volume types
+
+There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment.
```
How is this "default" volume type called? This should be standardized in SCS clouds.

anjastrunk replied:
I understand your point and we should add your explanations to the standard. IMO, the standard does not point out explicitly, that using no volume type, implies _default_ volume type is used, and the user can not expect any specific properties the default volume will have.

```suggestion
There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from default volume types.  
```

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+| Field              | Value                                                                                                                                                        |
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| access_project_ids | None                                                                                                                                                         |
+| description        | [encrypted][replicated] Content will be replicated three times to ensure consistency and availability for your data. LUKS encryption is used.                |
```
Please make the order of the keyword encrypted, replicated, number of replicas, ... in volume's description and name explicitly, similar to [flavor naming standard](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0100-v3-flavor-naming.md#complete-proposal-for-systematic-flavor-naming).

josephineSei replied:
1. I doubt that optional parameters like the number of replicas will ever be made a required standard. From my discussion with csps I think, they will stay optional. Therefore I would not like to include them into some keyword standard. But it should be enough if it is stated in the description.
2. That leaves us with only two aspects to standardize: [encrypted] and [replicated]. They both can occur together or alone. So possible beginnings for the description would be:
     - [encrypted]
     - [replicated]
     - [encrypted][replicated]
     - [replicated][encrypted]
3. We only recommend, that both required aspects are present, but not whether they should be combined or can be stand alone. That makes it harder to require a certain sequence.

So if you insist on having a special sequence I would add this, but I don't see any positive side effects from this.

anjastrunk replied:
The positive side effect of having a dedicated order of keywords is at user side, IMO. As volume type properties are currently defined as string, a cloud service consumer has to parse strings in order to find a volume type with special behavior or properties. Having a standardized order, string parsing will be facilitated.

artificial-intelligence replied:
@anjastrunk do you have a particular language in mind that relies on a certain order of arguments to facilitate string parsing?

imho most languages nowadays support some kind of simple pattern matching where the order is not that important?

overall I have no strong opinion if we should enforce the order, but we maybe should in either case explicitly define if this string is ordered or unordered.

fkr replied:
@anjastrunk please comment

mbuechse replied:
The text in the paragraphs above clearly states:

> The description needs to begin with `[encrypted]` ...

> The description needs to begin with `[replicated]` ...

Well, only one of these can be true. This does need some more work!

josephineSei replied:
I adjusted the document so that a special order is required now.

anjastrunk replied:
@fkr You are right, regular expression will not require any order. So I'm fine with extending the standard by adding, that key word are unordered.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+#### OPTIONAL addition: Number of Replicas
+
+Additionally to the fact, that a volume type is replicated, it OPTIONALLY can be stated in the description or name, how many times the provisioned volume will be replicated.
```
Please add an example.

josephineSei replied:
The example is right below, if you think, that is not sufficient i will add another one.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+## Conformance Tests
+
+As there are currently no REQUIRED volume types, we can only look for the RECOMMENDED aspects and thus executing a conformance test is not a must.
```
Now I'm confused. In line 44 you wrote, that there is always a default volume type. Doesn't this mean, that there is a required volume type?

josephineSei replied:
no. The default volume type is necessary through openstack. Like a server that always has a flavor, a volume always needs a volume type. But to be backward compatible the volume type must not needed to be specified. That requires the presence of a "default" volume type in openstack, but not in scs.

josephineSei commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+The following special terms are used throughout this standard document:
+| Term | Meaning |
+|---|---|
+| volume | OpenStack object, virtual drive which usually resides in a network storage backend |
```
```suggestion
| volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |
```

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+All Considerations can be looked up in detail in the [Decision Record for the Volume Type Standard.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. The following table shows, which aspects are considered in this standard and whether it is discoverable, and who can discover this. The comments state what can be done to help with the discoverability and other important restrictions.
```
```suggestion
To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. The following table shows, which aspects are considered in this standard. Furthermore the table shows, according to the current upstream implementation, whether the feature is discoverable, and who can discover this. The comments state what can be done to help with the discoverability and other important restrictions.
```

josephineSei commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,220 @@
+
+## DEFAULT volume types
+
+There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from default volume types.  
```
```suggestion
There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from default volume types.
```

akafazov reviewed (approved):
LGTM

mbuechse reviewed (commented):
From my POV, we should strictly separate between

- what volume types are mandatory/recommended
- how any volume type should be annotated to advertise its properties

I think this should be two separate sections.

Also, I think we should separate between

- what the standard mandates/recommends
- the whole discussion of the deficiencies of OpenStack and TODOs etc.

The CSP must be able to understand at one glance what they have to do; they may not want to read all the background info. Therefore the discussion should be put into a section of its own (design considerations).

artificial-intelligence commented:
> > The CSP must be able to understand at one glance what they have to do; they may not want to read all the background info. Therefore the discussion should be put into a section of its own (design considerations).
> 
> From this I get that I should put the "TODOs" I defined in the design consideration part - directly under the table. Would you agree with that?

fwiw I would like to say that I agree with this, as I think this makes the document easier to grasp for CSPs.

I still approved the change without this, as it's a nice to have for me, so not strictly necessary. Still it would be a good change imho.

mbuechse commented:
@josephineSei Maybe I was too quick. I'll try to be more thorough this time.

- The motivation uses the language "We want to". I don't understand that. We should mostly say what we DO, or even better, what the standard is about. Something very direct, like "This standard lists common features found in volume types, such as encryption, replication, availablity zones etc., and lays down how to advertise these features and make them discoverable."
- I would expect Design Considerations to be closer to [scs-0100-v3](https://github.com/SovereignCloudStack/standards/blob/feat/add_descriptions/Standards/scs-0100-v3-flavor-naming.md#complete-proposal-for-systematic-flavor-naming): one quick table of what features or information we consider, and a syntax scheme for how to encode them. BTW, I think it would be nice to follow the flavor-naming example and somehow have a kind of marker whether a certain volume type is intended to follow this scheme or not -- for the flavors, the name has to start with "SCS-", and we should find something similar here.
- Then I would continue with a section "Volume type features". Each feature should be described as succinctly as possible: what is necessary for a volume type to qualify as having this feature (but without command-line outputs or implementation details; these go into a separate section), and how do we make this feature discoverable (in line with the syntax scheme from above)? (From my point of view, this latter part should actually be binding!)
- Then I would expect a section "Recommended volume types", which mainly contains the table that you created in your previous comment.
- Finally, there can be sections with all the gory details about implementations, shortcomings of OpenStack, ideas for future developments etc.
- Also, the test script must of course check the syntax. It can't verify the properties, but the syntax must be checked, and this should not be marked TODO.

josephineSei commented:
> @josephineSei Maybe I was too quick. I'll try to be more thorough this time.
> 
>     * The motivation uses the language "We want to". I don't understand that. We should mostly say what we DO, or even better, what the standard is about. Something very direct, like "This standard lists common features found in volume types, such as encryption, replication, availablity zones etc., and lays down how to advertise these features and make them discoverable."

To me a motivation that only states what we already do is not a motivation. It then ceases to give me insights about the reason for the standard. What I considered should be in this section.
> 
>     * I would expect Design Considerations to be closer to [scs-0100-v3](https://github.com/SovereignCloudStack/standards/blob/feat/add_descriptions/Standards/scs-0100-v3-flavor-naming.md#complete-proposal-for-systematic-flavor-naming): one quick table of what features or information we consider, and a syntax scheme for how to encode them. BTW, I think it would be nice to follow the flavor-naming example and somehow have a kind of marker whether a certain volume type is intended to follow this scheme or not -- for the flavors, the name has to start with "SCS-", and we should find something similar here.

I disagree with the naming scheme. We will maybe eventually require 2 volume type aspects, but no more (we excluded them in the design phase). So we have the luxury of having enough space to write human understandable language. We should do so as we currently still MUST use the description for the Standard. (We would rather want to have discoverable options upstream....) 

The table for the aspects is present and detailed. I still added something to say what we want to standardize.

> 
>     * Then I would continue with a section "Volume type features". Each feature should be described as succinctly as possible: what is necessary for a volume type to qualify as having this feature (but without command-line outputs or implementation details; these go into a separate section), and how do we make this feature discoverable (in line with the syntax scheme from above)? (From my point of view, this latter part should actually be binding!)

I did this in the Decision document linked in the design decisions section. Imho we should not copy this.
> 
>     * Then I would expect a section "Recommended volume types", which mainly contains the table that you created in your previous comment.

done
> 
>     * Finally, there can be sections with all the gory details about implementations, shortcomings of OpenStack, ideas for future developments etc.

I don't quite understand that. It is already part of the subsections of the recommended part.
> 
>     * Also, the test script must of course check the syntax. It can't verify the properties, but the syntax must be checked, and this should not be marked TODO.

removed the todo

mbuechse commented:
This document is still too tedious for my taste, and it has become even more tedious with the most recent revision.

Imagine an engineer from a CSP who has a lot of these documents in front of them. In this case, it is a _standard_ titled _Volume type standard_. Then the first sentence under "Motivation" reads "We want to standardize a few varieties of volume types." This sentence is superfluous.

> To me a motivation that only states what we already do is not a motivation.

Well, let's not delve too deep into semantics here. The motivation of the document is to achieve a certain goal, and the goal, in this case, is to list certain features commonly found in volume types and make those discoverable. Let's not waste any more words on this than necessary. Think of the poor engineer.

> I disagree with the naming scheme.

Well, the current text already imposes a scheme, albeit one that is contradicting itself (see my other comment). It would be nice to make the requirements explicit and free of contradictions.

> We will maybe eventually require 2 volume type aspects

The text does describe many more aspects. Either we find a standardized way of describing them, or these aspects don't belong into the standard altogether.

> we currently still MUST use the description for the Standard

Yes. I have been talking about the description all along. My reference to the flavor-naming standard was not meant to imply that we standardize any names here, but that we find a similar syntax for (part of) the description.

I don't want to reiterate what I wrote earlier. The bottom line is that the text is too tedious for my taste.

mbuechse commented:
One more thing:

> I don't quite understand that. It is already part of the subsections of the recommended part.

But that is precisely my problem.

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,226 @@
+
+All Considerations can be looked up in detail in the [Decision Record for the Volume Type Standard.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. The following table shows, which aspects are considered in this standard. Furthermore the table shows, according to the current upstream implementation, whether the feature is discoverable, and who can discover this. The comments state what can be done to h ...
```
```suggestion
To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. The following table shows, which aspects are considered in this standard. Furthermore the table shows, according to the current upstream implementation, whether the feature is discoverable, and who can discover this. The comments state what can be done to help with the discoverability and other important restrictions. The last column says how the standard is handled, which is currently through the addition of a standardized phrase at the beginning of the description of a volume type.
```

artificial-intelligence commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,226 @@
+
+## REQUIRED volume types
+
+Currently the SCS will not require volume types with certain specification. This might change in the future.
```
```suggestion
Currently the SCS will not require volume types with certain specification. This will change in the future.
```

imho this will certainly change, no? otherwise, what's the point of the standard?

mbuechse commented:
If I may, a first draft proposal for a syntax would be along the lines of, for instance:

- `[scs:encrypted]`
- `[scs:replicated]`
- `[scs:encrypted,replicated]`

So, if the description starts with `[scs:`, then it has to contain a `]` as well; we take everything after the prefix until the first `]`, split at every comma, and we arrive at a list. This list must be ordered alphabetically, and the elements must be unique. We call this list the scs volume type descriptor.

So far -- the volume type descriptor only allows the elements `encrypted` and `replicated`. Now the question is why we don't want to encode other features as well, such as AZs.

josephineSei commented:
First question to everyone / maybe for a call: Is there a really good reason to put "scs" somehow / in some way into the standardized phrase? While it may come from the scs, why should it be stated everywhere, when it only makes the recognition of such phrases harder or less human readable?

2. We only have limited space in the description. See the [openstack code here](https://github.com/openstack/cinder/blob/a0d075219a1776f255e44982ffae652d5ab80506/cinder/db/sqlalchemy/models.py#L518) . While it does make sense to me to try it with a yaml/python-dict/something equivalent as identifier in front of the description, this might be too space consuming and makes it harder to be read by humans. So I would rather stay with tag-like identifiers.

3. As to why we don't want to do have more than these two aspects: Short version: Many aspects are either not widely used or are a mixture of settings in volume type AND storage backend AND compute backend  or are a mixture of those problems. Providing information at this single point that can easily be changed in more then one other place does not seem to be right. For detailed information please read the [decision record](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md). 

As mentioned in the other PR i could also remove any mention of OPTIONAL things. That might lighten the whole standard.

mbuechse commented:
Okay, so I'll try to explain some more. Say you have these prefixes of descriptions:

- `[encypted]`
- `[replicated][encrypted]`

Both of them are wrong from the POV of this standard: the first misses an `r`, and the second has the wrong order. But out script cannot tell whether it was ever intended to be 'right', can it? No, because our syntax forbids nothing.

However, in my proposals, it would be clear:

- `[scs:encypted]` and `scs: {encypted: true}` are plain wrong,
- `[scs:replicated,encrypted]` is wrong and `scs: {replicated: true, encrypted: true}` is actually ok (because with yaml, the order wouldn't matter)

artificial-intelligence commented:
> Okay, so I'll try to explain some more. Say you have these prefixes of descriptions:
> 
>     * `[encypted]`
> 
>     * `[replicated][encrypted]`
> 
> 
> Both of them are wrong from the POV of this standard: the first misses an `r`, and the second has the wrong order. But our script cannot tell whether it was ever intended to be 'right', can it? No, because our syntax forbids nothing.

Huh? you are contradicting yourself and the standard in just two sentences, so first, to quote the proposal:

> The standardized phrases are `[encrypted]` and `[replicated]`. They MUST appear at the beginning of the description of the volume type and they MUST be ordered alphabetically. Having only volume type number 1 would be enough to fulfill all recommendations. Having only volume types 2 and 3 together would also satisfy all recommendations

So at first you say "Both of them are wrong from the POV of this standard: the first misses an `r`[..]" and then you say: "[..]our syntax forbids nothing".

I don't understand this. it is clearly spelled out in the standard, that the standardized phrase is "encrypted" not "encypted", I would expect any test to catch that.

Am I somehow misunderstanding you? Because I can't find how you arrive at your conclusion that we can't check this. I fairly certain we can and should check the syntax?

> 
> However, in my proposals, it would be clear:
> 
>     * `[scs:encypted]` and `scs: {encypted: true}` are plain wrong,
> 
>     * `[scs:replicated,encrypted]` is wrong and `scs: {replicated: true, encrypted: true}` is actually ok (because with yaml, the order wouldn't matter)

The presence of a prefix does not say anything about the fact if a standard can be checked for syntax or not, it's just a prefix. This is completely orthogonal to the issue of being able to check syntax.

For the record: I have no strong opinion if we should or should not use a prefix, but it is an irrelevant property with regards to the question if a standards syntax can be parsed or not.

you can even parse plain strings with no ordering imposed, it is more work to parse, but it is possible.

mbuechse commented:
> I don't understand this. It is clearly spelled out in the standard, that the standardized phrase is "encrypted" not "encypted", I would expect any test to catch that.

Where does the standard say that `[encypted]` is forbidden?

Now I'm really losing my patience here. Yes, maybe my wording was not perfect. 

> Both of them are wrong from the POV of this standard

Of course, the standard does not and cannot label these as wrong (and what I meant in this quote is merely "mistaken"), but that is precisely the problem. We cannot catch these mistakes.

mbuechse commented:
The standard does not say anything about `[encypted]`. If anything, it says that `[encrypted]` MUST appear at the beginning, so if we are very strict, it is no longer possible to use this phrase anywhere else in the description, even if it also appears at the beginning? What kind of syntax is that? We need a clear delimiter of where the official syntax starts and ends, and then we can say what is correct and what isn't. Why is this so hard?

mbuechse reviewed (changes requested):
We need better syntax and the standard needs to be written more concisely.

**Edit**: and we need a test, for the standard refers to it ("the current test"), and it needs to be documented (at least the name of the script).

josephineSei commented:
The discussion in this PR seems futile. 

1. I strongly oppose to put anything like yaml/pythondict/etc into a string, that is used as a description. We should not abuse the description for things that should (and maybe eventually will) be handled in Upstream OpenStack. For now we need to work with what we have in a way that is clearly only a first version.
2.  I did not get any answers for my proposal to remove anything OPTIONAL. While we loop back to aspects, that are optional but imho (and in conclusion with some csps I spoke to for the decision record) should never be required. I originally added them for full transparency.

We should discuss all points left in a meeting, so possible misunderstandings and uncertainties can be addressed.

fkr commented:
thanks @josephineSei for raising this. Who do you (@josephineSei) think, should be part of that call?
@garloff is this something you could facilitate?

garloff commented:
Sure

garloff commented:
Sure.

josephineSei commented:
@fkr I think a discussion with @mbuechse and @artificial-intelligence would be good, also with maybe either you or @garloff  and maybe @anjastrunk.  Unfortunately with the holiday and vacation season incoming this will not take place before new year.

fkr commented:
> @fkr I think a discussion with @mbuechse and @artificial-intelligence would be good, also with maybe either you or @garloff and maybe @anjastrunk. Unfortunately with the holiday and vacation season incoming this will not take place before new year.

OK. I’ll make sure to set up that call once x-mas vacation period is over. @garloff will facilitate so that I can join the discussion.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+## Motivation
+
+We want to standardize a few varieties of volume types. While a user can choose simple things like size when creating a volume, Volume Types define a few broader aspects of a volume. Encryption of volumes for example is solely decided by the volume type. While the option with which the volume will be replicated is a mix between definition in the volume type and backend specific configuration, but ...
```
```suggestion
We want to standardize a few varieties of volume types. While a user can choose simple things like size when creating a volume, Volume Types define a few broader aspects of a volume. Encryption of volumes for example is solely decided by the volume type. While the option with which the volume will be replicated is a mix between definition in the volume type and backend specific configuration, but it's visibility can only be reached in the volume type. In the following part, we want to state, which varieties of volume types are REQUIRED, RECOMMENDED or only OPTIONAL within a SCS-compatible deployment.
```

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. The following table shows, which aspects are considered in this standard. Furthermore the table shows, according to the current upstream implementation, whether the feature is discoverable, and who can discover this. The comments state what can be done to h ...
+
+| Aspect | Part of Standard | Discoverability | comments | concluded standardized description |
```
Content of this table needs refinement:

- What does "Upstream work for ..." mean? Does this means, that upstream is currently working on this feature **or**   that this feature has to be implemented by someone else. Does "Upstream: setable (Typo?) ..." mean the same? 
- "backend-dependent" is too generic and does not answer the question of discoverability. A reference to more details would be helpful.
-  Discoverability: Does "yes" mean, that everybody can discover this property?

Just a minor request for change in favor of visibility: Either use lower-case upper case in column "Aspect" don't mix both.

josephineSei replied:
The table changed within the last commits.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+| Volume QoS | **Optional** | admin only | depends on the admin only volume qos object | - |
+| Backend name | no | - | - | - |
+
+In addition it is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but they don't have to be different volume types.
```
Typo? 

```suggestion
In addition it is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but there don't have to be different volume types.
```

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+In addition it is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but they don't have to be different volume types.
+Example: one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.
+
+The current downside is that users without admin rights currently cannot see the recommended aspects. We want and need to change this in the OpenStack workflow through adding discoverable properties (e.g. "encrypted"="true" and "replicated"="true") that are also visible for users. This way we will be able to automatically check whether a volume type with encyrption and or replication is present.  ...
```
I would not use first person plural. And I would not write, that SCS will extend OpenStack. I would just point out, that an extension in OpenStack is required, to support/fulfill this standard.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
````diff
@@ -0,0 +1,227 @@
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+We want to find a way to also use the internal extra_spec for replication, when the replication is automatically done by the backend. If this is not possible, we would like to introduce another property, which has to be set by the admin, when setting the volume type. Only after that we will have the possibility to automatically check for a volume type with replication.
````
This does not belong to this standard. It should be outsources in an own issue.

josephineSei replied:
done issue is #449

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+## DEFAULT volume types
+
+There is always a default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from default volume types.
```
```suggestion
There is always one default volume type defined in an OpenStack deployment. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from default volume types.
```

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+### Encryption
+
+Encryption for volumes is an option which has to be configured within the volume type. As an admin it is possible to set encryption-provider, key size, cipher and control location. And for admins it is also currently possible to see these configurations in a volume type with both list and show commands.
```
As this standard RECOMMENDs encrypted volumes, is MUST make recommendations at least for cipher and key-size as well. Cloud customers will buy a pig a poke otherwise.

artificial-intelligence replied:
well I just want to point out: we did debate this during one of the calls - I don't remember when exactly - and I also voiced concerns regarding this.

I can't remember the details (like "why"), but it was agreed upon to explicitly _not_ encode details about the used encryption scheme into the standard.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+### Replication
+
+Replication states, whether or not there are multiple replicas of a volume. Thus answers the question, whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:
```
In line 79, you define replicated volumes MUST be replicated three times. Here,  you dilute this requirement. This standard must be consistent and I recommend to add the number of replicas here as well.

artificial-intelligence replied:
as was discussed during a meeting regarding this standard:

There is currently no way to discover the replication factor used by the most used backend (ceph).

I believe this is even mentioned somewhere in the prelude of the standard, at least in some earlier iterations it was there.

Regardless, given the above facts, I don't think it makes sense to standardize a certain number of replicas, given that we can't test this.

josephineSei replied:
line 79 does only provide examples of possible volume types. Maybe this should be stated more clearly.

The number of replicas is difficult to get, because there are two ways replication can be achieved that exclude each other. 

1. Replication can be set in the volume type. There you can see the information that the volume will be replicated and the Number of Replicas OpenStack will set.
2. Performancewise it is better to let replication happen in the backend itself and not having the need of OpenStack caring about it. Ceph e.g. does it on its own. It is set in the configuration of ceph (https://docs.ceph.com/en/latest/rados/operations/pools/), where also the number of replicas can be set on a per pool base. So it is not possible for OpenStack to get the information (at least for now, but I do think this will be really hard to get into the volume type database of OpenStack.). The way we do have right now, is that the operator creating or editing the volume type with the ceph backend, will write down the information in the description field.

In conclusion I think the number of replicas would be really nice to have, but maybe will be too hard to get at all.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+### Availability Zones
+
+Availability Zones are not necessarily used by every deployer of a cloud. The SCS does not want do make it necessary to have AZs for a compatible deployment. But we want to encourage deployer who use multiple AZs to read the following:
```
I read "deployer" several time throughout this standard. I doubt if this role fits best. Other standards distinguish between provider and consumer, only.  If this standard uses "deployer", this role has to be introduced beforehand.

josephineSei replied:
maybe that is from my perspective as openstack-developer: I only see the interacting roles, and not the business roles. I can edit the standard, but would also like to agree on certain terms beforehand.

anjastrunk commented on Standards/scs-0112-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,227 @@
+
+### Volume QoS
+
+Quality of Service parameters for volumes can be defined in a volume qos object. While admins can use these qos objects directly on volumes, a user is not able to see such an object. Instead there can be ONE volume qos object associated to a volume type, which is then be used on all volumes created from that volume type. Through this indirection, volume types also working without these associated ...
```
>  While admins can use these qos objects directly on volumes

I do not understand this sentences. Can an admin define QoS objects for volumes directly? If so, how does the underlying infrastructure reflect this QoS? To reduce potential of confusion, I would remove this sentence and focus on user user only.


  ```suggestion
Quality of Service parameters for volumes can be defined in a volume QoSs object by admins. A user is not able to see such QoS objects. Instead there can be ONE volume QoS object associated to a volume type only, which is then be used on all volumes created from that volume type. Through this indirection, volume types also working without these associated QoS objects and this feature not being heavily used, the SCS standard will currently only view this as an OPTIONAL volume type feature.
```

josephineSei replied:
Okay so this seem to need further explanation:

`volume qos` is an extra ressource (or object) in openstack. Administrators can use it on volumes directly. Normal user ar not able to see these objects. So to use `volume qos` objects or QoS for volumes in general , an administrator has to create a `volume qos` object and associate it with a volume type (only one qos object per type). It is then applied to every volume created with this volume type. BUT still only admins are able to see the associated `volume qos` id in the volume type.

I think I will rephrase the fist sentences.

josephineSei replied:
optional aspects were removed from this standard - so this is no longer part of it.

josephineSei commented:
@fkr I created a doc for the meeting to group all points:
https://input.scs.community/o3_5fhYnS0eW14TcuprY1Q#

fkr commented:
@josephineSei thanks! that is really helpful. I suggest we use the overflow slot on Monday 22nd for that.

fkr commented:
https://github.com/SovereignCloudStack/calendar/pull/239

artificial-intelligence commented:
> > @fkr I think a discussion with @mbuechse and @artificial-intelligence would be good, also with maybe either you or @garloff and maybe @anjastrunk. Unfortunately with the holiday and vacation season incoming this will not take place before new year.
> 
> OK. I’ll make sure to set up that call once x-mas vacation period is over. @garloff will facilitate so that I can join the discussion.

I think we still need to do this meeting, or did I miss someting @fkr ? :slightly_smiling_face:

fkr commented:
> > > @fkr I think a discussion with @mbuechse and @artificial-intelligence would be good, also with maybe either you or @garloff and maybe @anjastrunk. Unfortunately with the holiday and vacation season incoming this will not take place before new year.
> > 
> > 
> > OK. I’ll make sure to set up that call once x-mas vacation period is over. @garloff will facilitate so that I can join the discussion.
> 
> I think we still need to do this meeting, or did I miss someting @fkr ? 🙂

yes, see my comment from an hour ago ;) It is now scheduled.

fkr commented:
We have to reschedule, since the required parties are not around today.
See: https://github.com/SovereignCloudStack/calendar/pull/242

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+## Introduction
+
+Volume Types are used to classify volumes and provide a basic decision for what kind of volume should be created. These volume types can sometimes be very backend-specific and it might be hard for a user to choose the most suitable volume type, if there is more than one type.
```
Comparing this MR with https://github.com/SovereignCloudStack/standards/pull/453/ I prefer MR #453, as it's more precise.

josephineSei replied:
I changed it.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+| AZ | Availability Zone |
+| Volume QoS | Quality of Service object for Volumes |
+
+## Motivation
```
I prefer motivation from #453

josephineSei replied:
I changed it.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type ...
+
+`[scs:aspect1][scs:aspect2]...`
```
IMO, we can omit repetition of prefix `scs`
```suggestion
`[scs:aspect1, aspect2,... aspect2]`
```

josephineSei replied:
i changed it.

anjastrunk commented:
LGTM.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+## DEFAULT volume type
+
+There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not exp ...
```
```suggestion
There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. This standard does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not expect any specific behavior or properties from volume types named `__default__`.
```

josephineSei replied:
i changed it.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. The SCS does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can not exp ...
+
+The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And the SCS will not make any assumptions about parameters being present in the default volume type.
```
```suggestion
The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.
```

josephineSei replied:
I changed it.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+## REQUIRED volume types
+
+Currently the SCS will not require volume types with certain specification. This will change in the future.
```
```suggestion
Currently, this standard will not require volume types with certain specification. This will change in the future.
```

josephineSei replied:
I changed it.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+
+## RECOMMENDED volume types
+
+The SCS recommends to have one or more volume types, that satisfy the need for encrpytion and replication.
```
```suggestion
This standard recommends to have one or more volume types, that satisfy the need for encrpytion and replication.
```

josephineSei replied:
I changed it.

josephineSei commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,128 @@
+| Field              | Value                                                                                                                                                        |
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| access_project_ids | None                                                                                                                                                         |
+| description        | [scs:encrypted][scs:replicated] Content will be replicated three times to ensure consistency and availability for your data. LUKS encryption is used.        |
```
i changed it.

josephineSei reviewed (commented):
I edited the standear as we discussed with @anjastrunk and @mbuechse

josephineSei commented:
I added and adjusted the tests from @mbuechse s proposal, so we have one unified PR.
I also tested the tests on my devstack:
```
stack@devstack:~/devstack$ openstack volume type list --long
+---------------------+-------------+-----------+---------------------+----------------------+
| ID                  | Name        | Is Public | Description         | Properties           |
+---------------------+-------------+-----------+---------------------+----------------------+
| 1ed91a96-6db1-4746- | test-type   | True      | None                |                      |
| b413-bfc42e3919c2   |             |           |                     |                      |
| d0ff021e-9379-452d- | LUKS3       | True      | [scs:encrypted]     | encryption_enabled=' |
| 8596-f1669447edbd   |             |           | blabla              | 1'                   |
| 08ca1605-26c0-4e50- | LUKS2       | True      | [scs:encrypted]     |                      |
| 8ffd-cb1b59ae42e4   |             |           | blabla              |                      |
| 39bc174e-c075-4296- | LUKS        | True      | [scs:encrypted]     | cat='miau'           |
| 8a09-f3b274ae0caa   |             |           | blabla              |                      |
| 75056530-6107-4bed- | ceph        | True      | [scs:replicated]    | volume_backend_name= |
| a473-f6ebe5f80ff1   |             |           | replicated          | 'ceph'               |
| afa6fbd5-0c4d-4928- | __DEFAULT__ | True      | Default Volume Type |                      |
| a50d-d18e49a35327   |             |           |                     |                      |
+---------------------+-------------+-----------+---------------------+----------------------+
stack@devstack:~/devstack$ python3 test-volume-types.py --os-cloud devstack --debug
DEBUG: Connecting to cloud 'devstack'
DEBUG: LUKS3: feature list ['encrypted', 'replicated']
DEBUG: LUKS2: feature list ['encrypted']
DEBUG: LUKS: feature list ['encrypted']
DEBUG: ceph: feature list ['replicated']
DEBUG: volume types by feature: {'encrypted': ['LUKS3', 'LUKS2', 'LUKS'], 'replicated': ['LUKS3', 'ceph']}
DEBUG: Total critical / error / warning: 0 / 0 / 0
```

It correctly identifies all volume types with one or multiple aspects and skips the one without a description.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+### Standardized Aspects
+
+The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:
```
I see the value of the additional column. The introductory paragraph is not very good though: the comma is wrong, and it's quite verbose. Why not use my proposal, which simply said

The following features are recognized.

Or if you must

The following features are recognized by this standard.

(But to my taste this is already a bit redundant; it's the point of the standard to state what's in the standard, so we need not repeat "in this standard" or "by this standard" all the time.)

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:
+
+| Aspect | Part of Standard | standardized description |
+| ---- | ---- | ------ |
+| Encryption | **Recommended** | **"[scs:encrypted]"** |
+| Replication | **Recommended** | **"[scs:replicated]"** |
```
I would still like to see the meaning column here.
The "Part of Standard" heading sounds quirky and needs to be renamed (or at the very least, make "standard" lower case)

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+| Encryption | **Recommended** | **"[scs:encrypted]"** |
+| Replication | **Recommended** | **"[scs:replicated]"** |
+
+It is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but there don't have to be different volume types.
+Example: one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.
```
```suggestion
It is possible that a single volume type combines multiple features.
For instance, one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+It is possible to use multiple of those aspects within one volume type. SCS will only ever look, if there is a volume type that has an aspect, but there don't have to be different volume types.
+Example: one volume type that uses LUKS-encryption with a ceph storage with inherent replication would fulfill all recommendations of this standard.
+
+## DEFAULT volume type
+
+There is always a default volume type defined in an OpenStack deployment. This volume type is created in the setup of cinder and will always be present in any OpenStack deployments under the name `__default__`. This standard does not have any requirements about this volume type at this moment, instead deployers are free to choose what fits best in their environment. Conversely, a cloud user can n ...
+
+The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.
```
I don't see the point in this: this assertion is entirely vapid. It's fine to say that no change of the default volume type is required by this standard, but this is only a side remark (because it follows from the text of the standard), and it warrants at most one sentence.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+The parameters of volume types described in this standard do not have to be applied to the chosen default volume type. And this standard will not make any assumptions about parameters being present in the default volume type.
+
+## REQUIRED volume types
+
+Currently, this standard will not require volume types with certain specification. This will change in the future.
```
I don't see the point in this. We always say that today's recommendations are tomorrow's requirements, and therefore all tests that we build MUST emit a warning, instead of a mere information, when some recommendation is violated. So this is vapid and can be dropped.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+### Encryption
+
+There SHOULD at least be one volume type with the encryption aspect.
```
This sentence is redundant. The rest of the section belongs under a new section "Implemenation notes" (which should be a Supplement).

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+## RECOMMENDED volume types
+
+This standard recommends to have one or more volume types, that satisfy the need for encrpytion and replication.
```
The comma is wrong. I prefer the more clear and easily testable assertions from my proposal:

At least one volume type SHOULD feature encryption.

At least one volume type SHOULD feature replicaton.

mbuechse replied:
One additional remark: 'satisfy the need for' -- satisfying needs is beyond this standard! We are talking about volume types having certain aspects or features. Whatever needs are satisfied by this (or not) is up the reader.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+### Replication
+
+There SHOULD at least be one volume type with the replication aspect.
```
This sentence is redundant. The rest of the section belongs under a new section "Implemenation notes" (which should be a Supplement).

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+1. In the configuration of a volume type. It then is visible as extra_spec in the properties of a volume type.
+2. Via the used backend. Ceph for example provides automatic replication, that does not need to be specified in the volume type. This is currently not visible for users.
+
+To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.
```
This whole sentence is again redundant. How many places of this document shall the reader check for inconsistencies before they can proceed with the implementation???

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.
+
+### Example
```
I like this, but again, this belongs under a section "Implementation notes".

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
````diff
@@ -0,0 +1,138 @@
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+## OPTIONAL volume types
+
+Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the CSPs, if they want to use some of these other aspects ...
````
Drop this paragraph. If it's unclear by now that other features are not recognized by this standard, and that arbitrary volume types may still exist, then we must improve the relevant places.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+## Related Documents
+
+[Here is the decision record document.](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)
```
```suggestion
- corresponding [decision record document](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0111-v1-volume-type-decisions.md)
```

don't say "here is" -- what does here even mean?

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+## Conformance Tests
+
+As there are currently no REQUIRED volume types, we can only look for the RECOMMENDED aspects and thus executing a conformance test is not a must.
+Furthermore the recommended aspects currently have to be described in the description by the deployer.
+Deployers MUST provided correct information, when adhering to this standard.
+The tests will only check the syntactical correctness.
+In future versions we aim to integrate some extra_specs for them in upstream OpenStack.
+It is also possible that a single volume type can currently fulfill all RECOMMENDED aspects.
+
+The current test will check for the presence of `[encrypted]` and `[replicated]` in the description of at least one volume type.
```
```suggestion
The script `/Tests/iaas/volume-types/volume-types-check.py` connects to an OpenStack environment and tests
the following:

- for each volume type: if its description starts with `[scs:....]`, then this prefix is a feature list
  (sorted, each entry at most once), and each entry is one of the possible features described here,
- the recommended volume types are present (otherwise, a WARNING is produced).

The return code is zero precisely when the test could be performed and the conditions are satisfied.
Otherwise, detailed errors and warnings are output to stderr.
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type ...
+
+`[scs:aspect1, aspect2, ..., aspectN]...`
```
@anjastrunk Is `scs` a bit short (and meaningless)? As far as I recall, the participants in the original discussion didn't like `scs`.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+`[scs:aspect1, aspect2, ..., aspectN]...`
+
+There is no sorting of aspects required. Every aspect should only be mentioned to the maximal amount of one.
```
Now we are back to no sorting? Why is that? As far as I recall, Anja suggested to use some sorting. I prefer sorting because it gives a canonical form which in the long run is easier to parse even for humans, because you don't have to read the whole thing to see if it has encryption (because, if it starts with replication, then it can't).

mbuechse replied:
Besides, you copied my script verbatim, and the script still checks whether the list is sorted.

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+|---|---|
+| volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |
+| volume feature | A certain feature a volume can possess |
+| aspect | part of a volume type that will activate a corresponding feature in a created volume |
```
```suggestion
| aspect | Part of a volume type that will activate a corresponding feature in a created volume |
```

anjastrunk commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+
+## OPTIONAL volume types
+
+Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the CSPs, if they want to use some of these other aspects ...
```
```suggestion
Any other aspects of volume types, that can be found in the decision record are OPTIONAL. They SHOULD NOT be referenced in the way this standard describes. Some of them already are natively discoverable by users, while others could be described in the name or description of a volume type. Users should look into the provided volume types of the Cloud Service Providers, if they want to use some of these other aspects.
```

anjastrunk commented on Tests/iaas/volume-types/volume-types-check.py:
```diff
@@ -0,0 +1,148 @@
+
+def print_usage(file=sys.stderr):
+    """Help output"""
+    print("""Usage: entropy-check.py [options]
```
print statement refers the wrong file, as it is called `volume-types-check.py`?

mbuechse replied:
```suggestion
    print("""Usage: volume-types-check.py [options]
```

anjastrunk reviewed (approved):
LGTM
Just, found three minor typos.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+The following special terms are used throughout this standard document:
+| Term | Meaning |
+|---|---|
+| volume | OpenStack ressource, virtual drive which usually resides in a network storage backend |
```
```suggestion
| volume | OpenStack resource, virtual drive which usually resides in a network storage backend |
```

mbuechse commented on Tests/iaas/volume-types/volume-types-check.py:
```diff
@@ -0,0 +1,148 @@
+    fs = match.group(1)
+    if not fs:
+        return []
+    return fs.split(", ")
```
```suggestion
    return [f.strip() for f in fs.split(",")]
```
I don't think we should mandate whitespace.

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+
+### Replication
+
+Replication states, whether or not there are multiple replicas of a volume. Thus answers the question, whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:
```
```suggestion
Replication states whether or not there are multiple replicas of a volume. Thus, it answers the question whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+
+### Encryption
+
+Encryption for volumes is an option which has to be configured within the volume type. As an admin it is possible to set encryption-provider, key size, cipher and control location. Additionally to be discoverable by users an admin has to edit the description and add `[scs:encrypted]` at the beginning or after another scs aspect. It should look like this example:
```
This is not quite true anymore:

> add `[scs:encrypted]` at the beginning or after another scs aspect.

```suggestion
Encryption for volumes is an option which has to be configured within the volume type. As an admin it is possible to set encryption-provider, key size, cipher and control location. Additionally to be discoverable by users, the description should start with an aspect list such as `[scs:encrypted]` (potentially with additional aspects). It should look like this example:
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+1. In the configuration of a volume type. It then is visible as extra_spec in the properties of a volume type.
+2. Via the used backend. Ceph for example provides automatic replication, that does not need to be specified in the volume type. This is currently not visible for users.
+
+To fulfill this recommentation for now, the admin needs to add `[scs:replicated]` at the beginning or after any other scs aspect into the descriotion of the volume type.
```
```suggestion
To fulfill this recommentation, the description should start with an aspect list such as  `[scs:replicated]` (potentially with additional aspects).
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+| Aspect | Requirement | standardized description | comment |
+| ---- | ---- | ------ | ------ |
+| Encryption | **Recommended** | **"[scs:encrypted]"** | used to encrypt the volume |
+| Replication | **Recommended** | **"[scs:replicated]"** | volume is replicated to aviod data loss in a case of hardware failure |
```
```suggestion
| Replication | **Recommended** | **"[scs:replicated]"** | volume is replicated to avoid data loss in a case of hardware failure |
```

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+
+| Aspect | Requirement | standardized description | comment |
+| ---- | ---- | ------ | ------ |
+| Encryption | **Recommended** | **"[scs:encrypted]"** | used to encrypt the volume |
```
```suggestion
| Encryption | **Recommended** | **"[scs:encrypted]"** | volume is encrypted |
```
to make it consistent with the other comment, and to avoid confusion about the meaning of "used to"  (what is used? by whom?)

mbuechse commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,139 @@
+
+### Standardized Aspects
+
+The following table shows, which aspects are considered in this standard. The last column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:
```
```suggestion
The following table shows which aspects are considered in this standard. The third column shows how the description of the volume type has to be adjusted, if the aspect is fulfilled:
```

mbuechse reviewed (approved):
I found a few typos and inconsistencies. I think they should be addressed, but I approve nonetheless so I don't block the process any further. Some of my remarks from earlier reviews still stand, particularly regarding language.

artificial-intelligence commented on Standards/scs-XXXX-v1-volume-type-standard.md:
```diff
@@ -0,0 +1,138 @@
+
+### Systematic Description of Volume Types
+
+To test whether a deployment has volume types with certain aspects, the discoverability of the parameters in the volume type has to be given. As for the time this standard is created, there is no way for users to discover all aspects through OpenStack commands. Therefore the aspects, that are fulfilled within a volume type, should be stated in the beginning of the **description** of a volume type ...
```
ack on `features` vs `aspects` - I like features more. but I would consider this a nit, as imho it's clear from the context that the same things are meant by these words.
It's also listed as synonymous in both merriam webster and oxford dictionary:

https://www.merriam-webster.com/thesaurus/aspect

artificial-intelligence reviewed (approved):
LGTM but all open comments should be addressed in any form before this is merged.

only slightly related to this change:
I also just noticed that seemingly all tests (so actual code) don't seem to have some distinct copyright/licence notice so would fall under the general licence used in this repository, which happens to be a CC licence.
In general this is a good licence but should imho not really be used for code.

I also raised the licence topic simultaneously with @mbuechse via Matrix Chat to put it onto the SIG Standardization Agenda.

In the meantime it might be worth to add an SPDX Licence Header to test.

fkr reviewed (approved):
Thanks @josephineSei for your work on this. 
Thanks everyone else who has worked on this!
